File: solutions/javascript/lasagna-master/1/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check

/**
 *
 * @param {number | undefined} remainingTime
 * @returns {string} the cooking status of the lasagna
 */
export const cookingStatus = (remainingTime) => {
    if (Number.isNaN(Number(remainingTime)))
        return "You forgot to set the timer.";
    if (remainingTime === 0) return "Lasagna is done.";
    return "Not done, please wait.";
};

/**
 *
 *  @param {Array<string>} layers
 *  @param {number} averagePrepTimePerLayer
 *  @returns {number | undefined}
 */

export const preparationTime = (layers, averagePrepTimePerLayer = 2) => {
    return layers.length * averagePrepTimePerLayer;
};

/**
 *
 *  @param {Array<string>} layers
 *  @returns {number | undefined}
 */

export const quantities = (layers) => {
    const test = ["noodles", "sauce"];

    const estimate = Object.create({ noodles: 0, sauce: 0 });
    for (const layer of layers) {
        if (test.includes(layer)) {
            const count = estimate[layer] ?? 0;
            estimate[layer] = count + 1;
        }
    }
    estimate.noodles *= 50;
    estimate.sauce *= 0.2;
    return estimate;
};

/**
 *
 *  @param {Array<string>} a
 *  @param {Array<string>} b
 *  @returns void
 */

export const addSecretIngredient = (a, b) => {
    b[b.length] = a[a.length - 1]
};


/**
 * @param {Record<string, number>} recipeObj
 * @param {number} portions
 *
 */

export const scaleRecipe = (recipeObj, portions) => {
    return Object.fromEntries(
        Object.entries(recipeObj).map(([k, v]) => [k, (v * portions) / 2]))
}

File: solutions/go/lasagna-master/1/lasagna_master.go
package lasagna

func PreparationTime(layers []string, minutesPerLayer int) int {
	if minutesPerLayer == 0 {
		minutesPerLayer = 2
	}
	return len(layers) * minutesPerLayer
}

func Quantities(layers []string) (int, float64) {
	var noodles int
	var sauce float64

	for _, layer := range layers {
		switch layer {
		case "noodles":
			noodles += 50
		case "sauce":
			sauce += 0.2
		}
	}

	return noodles, sauce
}

func AddSecretIngredient(friendsList []string, myList []string) {
	secretIngredient := friendsList[len(friendsList)-1]
	myList[len(myList)-1] = secretIngredient
}

func ScaleRecipe(quantities []float64, portions int) []float64 {
	scaled := make([]float64, len(quantities))

	for i, quantity := range quantities {
		scaled[i] = quantity / 2.0 * float64(portions)
	}

	return scaled
}
